### ISSUE TITLE
Validator in STI parent model, but child throws MissingRequiredValidatorError

### ISSUE THREAD

**Issue by davidbenton** [thoughtbot/paperclip#1969]
I have a set of models using single-table inheritance with an image attachment (and the required validation) on the parent class. The code was working correctly, but on upgrading to current versions of Rails and Paperclip it began to throw `MissingRequiredValidatorError`.

The issue, if you can call it that, is that `ChildModel.validators` doesn't list validators from the parent model. I'm not sure if that should be considered a bug in Rails, Paperclip, or none of the above, but I wanted to raise the issue here. Thoughts?

For now, I've worked around this by adding `do_not_validate_attachment_file_type` to my child models as necessary. Hopefully this is temporary, as it's not the functionality I want.

**Comment by davidbenton** [comment]
FYI, here is Paperclip code that checks a model's valdators https://github.com/thoughtbot/paperclip/blob/19a3bfc6dc286fb9695cf48fb09f48de7f711435/lib/paperclip/attachment.rb#L391

**Comment by jyurek** [comment]
What version of Rails are you using? i just attempted this with 4.1.2 with the [paperclip demo project](https://github.com/thoughtbot/paperclip_demo), and it seems to not be working as you describe. I added an `OldFriend` model that inherited from `Friend`, and in the console, `OldFriend.validators == Friend.validators` was true.

**Comment by davidbenton** [comment]
Hi Jon, thanks for taking a look. I am using Rails 4.2.3.

Did you add validators to `Friend`? Here are my models, simplified:

```
# social_post.rb
class SocialPost < ActiveRecord::Base
  require_dependency 'social_post/facebook'

  self.inheritance_column = :type
  validates_presence_of :type # no instantiation of this (base) class
  validates_presence_of :datetime

  has_attached_file :image
  validates_attachment :image, content_type: { content_type: ["image/jpg", "image/jpeg", "image/png", "image/gif"] }

  # yadda yadda

end
```

```
# social_post/facebook.rb

class SocialPost::Facebook < SocialPost
  
  # does things – no attachments or validations defined in this child model

end
```

And here's what I get in Rails console (validators in parent, none in child):
```
[1] pry(main)> SocialPost.validators
=> [#<ActiveRecord::Validations::PresenceValidator:0x007fa33ccb43f0 @attributes=[:type], @options={}>,
 #<ActiveRecord::Validations::PresenceValidator:0x007fa33ccae6f8 @attributes=[:datetime], @options={}>,
 #<ActiveModel::BlockValidator:0x007fa33ccac740
  @attributes=[:image],
  @block=#<Proc:0x007fa33ccac880@/Users/dbenton/.rbenv/versions/2.1.3/lib/ruby/gems/2.1.0/gems/paperclip-4.3.0/lib/paperclip/has_attached_file.rb:27>,
  @options={}>,
 #<Paperclip::Validators::MediaTypeSpoofDetectionValidator:0x007fa33cc85258
  @attributes=[:image],
  @options={:if=>#<Proc:0x007fa33cc855f0@/Users/dbenton/.rbenv/versions/2.1.3/lib/ruby/gems/2.1.0/gems/paperclip-4.3.0/lib/paperclip/has_attached_file.rb:86 (lambda)>}>,
 #<Paperclip::Validators::AttachmentContentTypeValidator:0x007fa33cc74250
  @attributes=[:image],
  @options={:content_type=>["image/jpg", "image/jpeg", "image/png", "image/gif"], :allow_nil=>true}>]
[2] pry(main)> SocialPost::Facebook.validators
=> []
```

**Comment by davidbenton** [comment]
Note: My validations on the child model still work, despite `SocialPost::Facebook.validators` returning no validators (i.e. `[]`).

**Comment by davidbenton** [comment]
tl;dr My take is that if this is an issue at all, it's with Rails. 

I worked through this using the paperclip demo project to make a test case, and I have identified the source of the issue, though not yet the reason. Adding `require_dependency` as in the parent model above...
```
require_dependency 'social_post/facebook'
```

...causes the child class to no longer return the validators of the parent class (though validations continue to work as expected).

I added those requires because I was getting errors dynamically calling child classes from the parent without them. I will dig further, but if this is an issue, I would deem it to be a Rails issue, as the result of `ChildModel.validators` seems unintuitive. It's also possible there's a good Ruby or Rails explanation for why this works as it does.

**Comment by brendon** [comment]
Maybe this is a Rails bug, or perhaps it was a documented change? I've had a bit of a look but can't currently find any reference to it. Obviously Rails isn't relying on the `validators` method for actually running the validations or they wouldn't run.

Perhaps there's a case for at least working around this problem in Paperclip? Otherwise, without monkey patching, the only solution is to repeatedly add the validations to each STI class.

**Comment by brendon** [comment]
@davidbenton, were you perhaps using engines to store these particular classes? I'm trying to replicate the problem with a simple rails bug report test but it doesn't seem to manifest. Perhaps it's to do with how the classes are loaded?

**Comment by brendon** [comment]
@davidbenton, I figured out my problem. I'd changed my superclass to literally mention all of the subclasses (assigning them to a hash constant in the superclass). This killed things basically because of circular dependency. I'll probably look to turn the constant into a class method instead so it's not evaluated when the class is loaded.
